demonstrations/tutorial_ml_classical_shadows.py

```diff
@@ -0,0 +1,867 @@
+        for i, j in it.product(range(num_spins), range(num_spins)):
+            if not coupling_mats[itr][i][j]:
+                if (j % num_cols and j - i == 1) or (j - i == num_cols):
+                    coupling_mats[itr][i][j] = next(coup_terms)
+                    coupling_mats[itr][j][i] = coupling_mats[itr][i][j]
+    return coupling_mats
+
+
+######################################################################
+# For this demo, we study the model with four spins arranged on the nodes of
+# a square lattice that would require four qubits for the simulation, i.e.,
+# one qubit for one spin each. For constructing an instance of this model,
+# we first build the coupling matrix using our previously defined function.
+#
+
+Nr, Nc = 2, 2
+num_qubits = Nr * Nc  # Ns
+J_mat = build_coupling_mats(1, Nr, Nc)[0]
+
+
+######################################################################
+# We can visualize the model instance by representing the coupling matrix as a
+# ``networkx`` graph:
+
+import matplotlib.pyplot as plt
+import networkx as nx
+
+G = nx.from_numpy_matrix(np.matrix(J_mat), create_using=nx.DiGraph)
+pos = {i: (i % Nc, -(i // Nc)) for i in G.nodes()}
+edge_labels = {(x, y): np.round(J_mat[x, y], 2) for x, y in G.edges()}
+weights = [x + 1.5 for x in list(nx.get_edge_attributes(G, "weight").values())]
+
+nx.draw(
+    G,
+    pos,
+    node_color="lightblue",
+    with_labels=True,
+    node_size=600,
+    width=weights,
+    edge_color="firebrick",
+)
+nx.draw_networkx_edge_labels(G, pos=pos, edge_labels=edge_labels)
+plt.show()
+
+
+######################################################################
+# We then use the same coupling matrix :math:`J_mat` to obtain the Hamiltonian
+# :math:`H` for the model we have instantiated above. 
+#
+
+import pennylane as qml
+
+def Hamiltonian(J_mat):
+    coeffs, ops = [], []
+    for i in range(J_mat.shape[0]):
+        for j in range(i + 1, J_mat.shape[0]):
+            for op in [qml.PauliX, qml.PauliY, qml.PauliZ]:
+                coeff = J_mat[i, j]
+                if coeff:
+                    coeffs.append(coeff)
+                    ops.append(op(i) @ op(j))
+    H = qml.Hamiltonian(coeffs, ops)
+    return H
+
+print(Hamiltonian(J_mat))
+
+
+######################################################################
+# For the Heisenberg model, a propetry of interest is usally the two-body
+# correlation function :math:`C_{ij}`, which for a pair of spins :math:`\sigma^{z}_{i}`
+# and :math:`\sigma^{z}_{j}` is defined as:
+#
+# .. math::  \hat{C}_{ij} = \frac{1}{3} (X_i X_j + Y_iY_j + Z_iZ_j)
+#
+# Expectation value of each such element :math:`\hat{C}_{ij}` with respect to
+# the ground state :math:`|\psi_{0}\rangle` of the model can be used to build
+# the correlation matrix :math:`C`, such that:
+#
+# .. math:: {C}_{ij} = \langle \hat{C}_{ij} \rangle = \frac{1}{3} \langle \psi_{0} | X_i X_j + Y_iY_j + Z_iZ_j | \psi_{0} \rangle
+#
+
+
+def corr_function_op(i, j):
```

Comment:
```suggestion
def corr_function(i, j):
```
Or something else, but it's best to keep the number of underscores to a minimum in function and variable naming